apiserver/facades/agent/caasapplication/state.go

```diff
@@ -0,0 +1,115 @@
+// Copyright 2020 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package caasapplication
+
+import (
+	"github.com/juju/charm/v8"
+	"github.com/juju/names/v4"
+	"github.com/juju/version"
+
+	jujucontroller "github.com/juju/juju/controller"
+	"github.com/juju/juju/core/network"
+	"github.com/juju/juju/state"
+)
+
+// State provides the subset of global state
+// required by the CAAS application facade.
+type State interface {
+	Application(string) (Application, error)
+	Model() (Model, error)
+	Unit(name string) (Unit, error)
+	ControllerConfig() (jujucontroller.Config, error)
+	APIHostPortsForAgents() ([]network.SpaceHostPorts, error)
+}
+
+// Model provides the subset of CAAS model state required
+// by the CAAS application facade.
+type Model interface {
+	Containers(providerIds ...string) ([]state.CloudContainer, error)
+	AgentVersion() (version.Number, error)
+	ControllerTag() names.ControllerTag
+	Tag() names.Tag
+}
+
+// Application provides the subset of application state
+// required by the CAAS application facade.
+type Application interface {
+	Life() state.Life
+	Charm() (Charm, bool, error)
+	Name() string
+	AllUnits() ([]Unit, error)
+	UpdateUnits(unitsOp *state.UpdateUnitsOperation) error
+	AddUnit(args state.AddUnitParams) (unit Unit, err error)
+}
+
+// Charm provides the subset of charm state required by the
+// CAAS application facade.
+type Charm interface {
+	Meta() *charm.Meta
+}
+
+type stateShim struct {
+	*state.State
+}
+
+func (s stateShim) Application(id string) (Application, error) {
+	app, err := s.State.Application(id)
+	if err != nil {
+		return nil, err
+	}
+	return applicationShim{app}, nil
+}
+
+func (s stateShim) Model() (Model, error) {
+	model, err := s.State.Model()
+	if err != nil {
+		return nil, err
```

Comment:
Fix all the bare errors being sent back!

```suggestion
		return nil, errors.Trace(err)
```